File: modules/hashtables/hash_function_comparison.cpp
// Hash Function Quality Comparison  (CSC310 Module 4B - Hashing)
// =============================================================
//
// Measures (1) distribution: collisions/empty/longest bucket when keys are
// hashed into a prime-sized table, and (2) avalanche: when one input bit flips,
// a good hash flips about half the output bits.
//
// Build:  g++ -std=c++17 -O2 hash_function_comparison.cpp -o hash_function_comparison && ./hash_function_comparison
#include <cmath>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::string;
using std::vector;

const double A = 0.6180339887; // (sqrt(5)-1)/2

int hDivision(long k, int m)       { return (int)(k % m); }
int hMultiplication(long k, int m) { double f = (k * A); f -= (long)f; return (int)(m * f); }
int hShiftFold(long k, int m) {
    string s = std::to_string(k);
    long total = 0;
    for (size_t i = 0; i < s.size(); i += 2) total += std::stol(s.substr(i, 2));
    return (int)(total % m);
}
int hMidSquare(long k, int m) {
    string s = std::to_string((long long)k * k);
    size_t mid = s.size() / 2;
    string chunk = s.substr(mid > 0 ? mid - 1 : 0, 3);
    return (int)(std::stol(chunk) % m);
}
int hFirstDigit(long k, int m) { return (std::to_string(k)[0] - '0') % m; }

// ---- 32-bit hashes for the avalanche test ----
uint32_t mix32(uint32_t k) {       // good
    k *= 2654435761u; k ^= k >> 16;
    k *= 2246822519u; k ^= k >> 13;
    return k;
}
uint32_t weak32(uint32_t k) { return k * 3u; }  // poor: low bits only

double avalanche(uint32_t (*h)(uint32_t), int trials = 4000) {
    std::mt19937 rng(0);
    long flips = 0;
    for (int t = 0; t < trials; ++t) {
        uint32_t k = rng(), base = h(k);
        int bit = rng() % 32;
        uint32_t diff = base ^ h(k ^ (1u << bit));
        flips += __builtin_popcount(diff);
    }
    return (double)flips / trials / 32.0;
}

int main() {
    std::mt19937 rng(7);
    std::uniform_int_distribution<long> keygen(100000, 999999);
    vector<long> keys;
    for (int i = 0; i < 80; ++i) keys.push_back(keygen(rng));
    int m = 97, n = (int)keys.size();
    double expected = n - m * (1 - std::pow((m - 1.0) / m, n));

    struct S { const char* name; int (*h)(long, int); } strat[] = {
        {"division", hDivision}, {"multiplication", hMultiplication},
        {"shift folding", hShiftFold}, {"mid-square", hMidSquare},
        {"first digit (poor)", hFirstDigit},
    };

    std::cout << "Distribution of " << n << " keys into a table of size " << m
              << " (ideal uniform hash collides ~" << (int)std::lround(expected) << " times):\n";
    std::cout << "  " << std::left << std::setw(20) << "strategy" << std::right
              << std::setw(6) << "used" << std::setw(7) << "empty"
              << std::setw(10) << "max load" << std::setw(12) << "collisions" << "\n";
    std::cout << "  " << string(55, '-') << "\n";
    for (auto& s : strat) {
        vector<int> b(m, 0);
        for (long k : keys) b[s.h(k, m)]++;
        int used = 0, mx = 0;
        for (int x : b) { if (x) used++; mx = std::max(mx, x); }
        std::cout << "  " << std::left << std::setw(20) << s.name << std::right
                  << std::setw(6) << used << std::setw(7) << (m - used)
                  << std::setw(10) << mx << std::setw(12) << (n - used) << "\n";
    }
    std::cout << "\nAvalanche (fraction of 32 output bits that flip per input-bit flip;"
                 "\na good hash is near 0.50):\n";
    std::cout << "  mix32 (good multiply + xorshift): " << std::fixed
              << std::setprecision(3) << avalanche(mix32) << "\n";
    std::cout << "  weak32 (low bits only)          : " << avalanche(weak32) << "\n";
    return 0;
}

File: modules/hashtables/hash_function_comparison.py
"""
Hash Function Quality Comparison  (CSC310 Module 4B - Hashing)
============================================================

Two quality measures from the notes:
  - Distribution: collisions / empty slots / longest bucket when n keys are
    hashed into a prime-sized table.  A good hash spreads keys evenly.
  - Avalanche: when one input bit changes, about half the output bits should
    flip.  A weak hash barely changes its output.

Strategies (division, multiplication, folding, mid-square) are from the notes'
"Hashing Strategies" section; "first digit" is the deliberately poor one.
"""
import random
import statistics


# ---------- table-index strategies: map integer key -> [0, m) ----------
A = 0.6180339887  # (sqrt(5)-1)/2, Knuth's multiplicative constant

def h_division(k, m):
    return k % m

def h_multiplication(k, m):
    return int(m * ((k * A) % 1.0))

def h_shift_fold(k, m):
    s = str(k)
    total = sum(int(s[i:i + 2]) for i in range(0, len(s), 2))  # 2-digit groups
    return total % m

def h_mid_square(k, m):
    sq = str(k * k)
    mid = sq[len(sq) // 2 - 1: len(sq) // 2 + 2] or sq      # middle ~3 digits
    return int(mid) % m

def h_first_digit(k, m):  # deliberately poor: only the leading digit
    return int(str(k)[0]) % m


STRATEGIES = [
    ("division", h_division),
    ("multiplication", h_multiplication),
    ("shift folding", h_shift_fold),
    ("mid-square", h_mid_square),
    ("first digit (poor)", h_first_digit),
]


def distribution_report(keys, m):
    # birthday-style expected collisions for an ideal uniform hash, as a baseline
    n = len(keys)
    expected = n - m * (1 - ((m - 1) / m) ** n)
    print(f"Distribution of {n} keys into a table of size {m} "
          f"(ideal uniform hash collides ~{expected:.0f} times):")
    print(f"  {'strategy':<20}{'used':>6}{'empty':>7}{'max load':>10}{'collisions':>12}")
    print("  " + "-" * 55)
    for name, h in STRATEGIES:
        buckets = [0] * m
        for k in keys:
            buckets[h(k, m)] += 1
        used = sum(1 for b in buckets if b)
        print(f"  {name:<20}{used:>6}{m - used:>7}{max(buckets):>10}"
              f"{len(keys) - used:>12}")


# ---------- 32-bit hashes for the avalanche test ----------
MASK = 0xFFFFFFFF

def mix32(k):                       # good: multiply + xorshift (avalanches well)
    k = (k * 2654435761) & MASK
    k ^= k >> 16
    k = (k * 2246822519) & MASK
    k ^= k >> 13
    return k & MASK

def weak32(k):                      # poor: keep low bits only
    return (k * 3) & MASK


def avalanche(hfun, trials=4000):
    """Average fraction of the 32 output bits that flip when one input bit flips."""
    rng = random.Random(0)
    total_flip = 0
    for _ in range(trials):
        k = rng.getrandbits(32)
        base = hfun(k)
        bit = rng.randrange(32)
        diff = base ^ hfun(k ^ (1 << bit))
        total_flip += bin(diff).count("1")
    return total_flip / trials / 32.0


def main():
    rng = random.Random(7)
    keys = [rng.randint(100000, 999999) for _ in range(80)]   # 80 6-digit keys
    m = 97                                                    # prime table size

    distribution_report(keys, m)
    print()
    print("Avalanche (fraction of 32 output bits that flip per input-bit flip;")
    print("a good hash is near 0.50):")
    print(f"  mix32 (good multiply + xorshift): {avalanche(mix32):.3f}")
    print(f"  weak32 (low bits only)          : {avalanche(weak32):.3f}")


if __name__ == "__main__":
    main()
